trollmoves/tests/test_dispatcher.py

```diff
@@ -397,17 +399,69 @@ def test_create_dest_url():
             # SSH protocol, no username
-            url, params = dp.create_dest_url(msg, 'target2', config['target2'])
+            url, params, client = dp.create_dest_url(msg, 'target2',
+                                                     config['target2'])
             expected_url = "ssh://server.target2.com/satellite/viirs/sat_201909190919_NOAA-20.tif"
             assert url == expected_url
             assert params == {'ssh_key_filename': '~/.ssh/rsa_id.pub'}
+            assert client == "target2"
 
             # SCP protocolw with username
-            url, params = dp.create_dest_url(msg, 'target3', config['target3'])
+            url, params, client = dp.create_dest_url(msg, 'target3',
+                                                     config['target3'])
             expected_url = "scp://user@server.target2.com/satellite/viirs/sat_201909190919_NOAA-20.tif"
             assert url == expected_url
+            assert client == "target3"
 
     finally:
         if dp is not None:
             dp.close()
         os.remove(config_file_name)
+
+
+test_yaml_pub = test_yaml2 + """
+target3:
+  host: scp://user@server.target2.com
+  filepattern: 'sat_{start_time:%Y%m%d%H%M}_{platform_name}.{format}'
+  directory: /satellite/{sensor}
+  publish_topic: "/topic/{platform_name}"
+  dispatch_configs:
+    - topics:
+        - /level2/viirs
+"""
+@patch('trollmoves.dispatcher.Message')
+@patch('trollmoves.dispatcher.ListenerContainer')
+@patch('trollmoves.dispatcher.NoisyPublisher')
+def test_publisher(NoisyPublisher, ListenerContainer, Message):
+    """Test the publisher is initialized."""
+    pub = Mock()
+    NoisyPublisher.return_value = pub
+    with NamedTemporaryFile('w', delete=False) as config_file:
+        config_file_name = config_file.name
+        config_file.write(test_yaml_pub)
+        config_file.flush()
+        config_file.close()
+    try:
+        dp = Dispatcher(config_file_name, publish_port=40000,
+                        publish_nameservers=["asd"])
+
+        assert dp.publisher is pub
+        init_call = call("dispatcher", port=40000, nameservers=["asd"])
+        assert init_call in NoisyPublisher.mock_calls
+
+        msg = Mock(data={'uri': 'original_path',
+                         'platform_name': 'platform'})
+        destinations = [['url1', 'params1', 'target2'],
+                        ['url2', 'params2', 'target3']]
+        success = {'target2': False, 'target3': True}
+        dp._publish(msg, destinations, success)
+        dp.publisher.send.assert_called_once()
+        # The message topic has been composed and uri has been replaced
+        msg_call = call('/topic/platform', 'file',
+                        {'uri': 'url2', 'platform_name': 'platform'})
+        assert msg_call in Message.mock_calls
+    finally:
+        os.remove(config_file_name)
+        if dp is not None:
+            dp.close()
+            dp.publisher.stop.assert_called()
```

Comment:
Why not push all this one step in and let the NamedTemporaryFile delete the file ?

Reply:
Done.